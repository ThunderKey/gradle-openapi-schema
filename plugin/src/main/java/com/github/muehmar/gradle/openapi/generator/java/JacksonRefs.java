package com.github.muehmar.gradle.openapi.generator.java;

public class JacksonRefs {
  private JacksonRefs() {}

  public static final String JSON_POJO_BUILDER =
      "com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder";
}
