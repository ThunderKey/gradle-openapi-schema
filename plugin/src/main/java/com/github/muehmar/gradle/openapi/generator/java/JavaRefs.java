package com.github.muehmar.gradle.openapi.generator.java;

public class JavaRefs {
  private JavaRefs() {}

  public static final String LOCAL_DATE = "java.time.LocalDate";
  public static final String OPTIONAL = "java.util.Optional";
}
