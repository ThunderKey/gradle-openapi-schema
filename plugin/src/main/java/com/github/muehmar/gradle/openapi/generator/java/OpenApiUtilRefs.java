package com.github.muehmar.gradle.openapi.generator.java;

public class OpenApiUtilRefs {
  private OpenApiUtilRefs() {}

  public static final String OPENAPI_UTIL_PACKAGE = "com.github.muehmar.openapi.util";

  public static final String TRISTATE = OPENAPI_UTIL_PACKAGE + ".Tristate";
  public static final String JACKSON_NULL_CONTAINER =
      OPENAPI_UTIL_PACKAGE + ".JacksonNullContainer";
}
