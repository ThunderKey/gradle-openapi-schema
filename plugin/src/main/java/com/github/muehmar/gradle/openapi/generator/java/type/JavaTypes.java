package com.github.muehmar.gradle.openapi.generator.java.type;

public class JavaTypes {
  private JavaTypes() {}

  public static final JavaType OBJECT = JavaType.ofName("Object");

  public static final JavaType BOOLEAN = JavaType.ofName("Boolean");
  public static final JavaType BYTE = JavaType.ofName("Byte");
  public static final JavaType STRING = JavaType.ofName("String");
  public static final JavaType INTEGER = JavaType.ofName("Integer");
  public static final JavaType LONG = JavaType.ofName("Long");
  public static final JavaType FLOAT = JavaType.ofName("Float");
  public static final JavaType DOUBLE = JavaType.ofName("Double");
  public static final JavaType BYTE_ARRAY = JavaType.ofName("byte[]");

  public static final JavaType LOCAL_DATE =
      JavaType.ofNameAndImport("LocalDate", "java.time.LocalDate");
  public static final JavaType LOCAL_TIME =
      JavaType.ofNameAndImport("LocalTime", "java.time.LocalTime");
  public static final JavaType LOCAL_DATE_TIME =
      JavaType.ofNameAndImport("LocalDateTime", "java.time.LocalDateTime");

  public static final JavaType URI = JavaType.ofNameAndImport("URI", "java.net.URI");
  public static final JavaType URL = JavaType.ofNameAndImport("URL", "java.net.URL");

  public static final JavaType UUID = JavaType.ofNameAndImport("UUID", "java.util.UUID");
}
