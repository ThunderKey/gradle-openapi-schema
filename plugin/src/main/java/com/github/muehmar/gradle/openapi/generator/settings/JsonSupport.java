package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import java.io.Serializable;
import java.util.Optional;

public enum JsonSupport implements Serializable {
  JACKSON("jackson"),
  NONE("none");

  private final String value;

  JsonSupport(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<JsonSupport> fromString(String value) {
    return PList.of(values()).find(support -> support.value.equalsIgnoreCase(value));
  }
}
