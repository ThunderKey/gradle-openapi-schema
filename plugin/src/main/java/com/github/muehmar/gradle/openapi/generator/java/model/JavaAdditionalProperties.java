package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaIdentifier MAP_PROPERTY_NAME =
      JavaIdentifier.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, JavaAnyType.create());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, JavaAnyType.create());
  }

  public static JavaIdentifier getPropertyName() {
    return MAP_PROPERTY_NAME;
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaIdentifier.fromName(type.getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
