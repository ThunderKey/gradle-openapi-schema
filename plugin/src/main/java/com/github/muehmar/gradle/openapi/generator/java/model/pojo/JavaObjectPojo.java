package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ConstructorContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaIdentifier;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.wrap(type.mapName(objectPojo.getName())),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(comp -> JavaOneOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(name.getSchemaName());
  }

  @Override
  public JavaIdentifier getClassName() {
    return name.asJavaName().asIdentifier();
  }

  public JavaIdentifier prefixedClassNameForMethod(String prefix) {
    if (prefix.isEmpty()) {
      return name.asJavaName().startLowerCase().asIdentifier();
    } else {
      return name.asJavaName().startUpperCase().prefix(prefix).asIdentifier();
    }
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  private PList<JavaPojoMember> getAllOfMembers() {
    return allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getOneOfMembers() {
    return oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getAnyOfMembers() {
    return anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    return getAllOfMembers()
        .concat(getOneOfMembers())
        .concat(getAnyOfMembers())
        .distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return MemberContentBuilder.create()
        .isArrayPojo(false)
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return HashCodeContentBuilder.create()
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return EqualsContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return ToStringContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return ConstructorContentBuilder.create()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return WitherContentBuilder.create()
        .className(getClassName())
        .membersForWithers(getMembers())
        .membersForConstructorCall(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
