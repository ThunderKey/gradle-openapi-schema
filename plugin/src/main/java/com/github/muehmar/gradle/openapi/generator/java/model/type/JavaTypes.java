package com.github.muehmar.gradle.openapi.generator.java.model.type;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.type.ArrayType;
import com.github.muehmar.gradle.openapi.generator.model.type.EnumType;
import com.github.muehmar.gradle.openapi.generator.model.type.IntegerType;
import com.github.muehmar.gradle.openapi.generator.model.type.MapType;
import com.github.muehmar.gradle.openapi.generator.model.type.ObjectType;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;

public class JavaTypes {
  private JavaTypes() {}

  public static final JavaArrayType STRING_LIST =
      JavaArrayType.wrap(ArrayType.ofItemType(StringType.noFormat()), TypeMappings.empty());
  public static final JavaBooleanType BOOLEAN = JavaBooleanType.wrap(TypeMappings.empty());
  public static final JavaEnumType ENUM =
      JavaEnumType.wrap(
          EnumType.ofNameAndMembers(
              Name.ofString("Gender"), PList.of("male", "female", "divers", "other")));
  public static final JavaMapType MAP =
      JavaMapType.wrap(
          MapType.ofKeyAndValueType(StringType.noFormat(), StringType.noFormat()),
          TypeMappings.empty());
  public static final JavaNoType NO_TYPE = JavaNoType.create();
  public static final JavaIntegerType INTEGER =
      JavaIntegerType.wrap(IntegerType.formatInteger(), TypeMappings.empty());
  public static final JavaObjectType OBJECT =
      JavaObjectType.wrap(ObjectType.ofName(PojoName.ofName(Name.ofString("UserDto"))));
  public static final JavaStringType STRING =
      JavaStringType.wrap(StringType.noFormat(), TypeMappings.empty());
}
