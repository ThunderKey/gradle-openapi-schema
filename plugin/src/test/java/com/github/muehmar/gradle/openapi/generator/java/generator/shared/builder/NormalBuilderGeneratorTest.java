package com.github.muehmar.gradle.openapi.generator.java.generator.shared.builder;

import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JacksonRefs;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.OpenApiUtilRefs;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes;
import com.github.muehmar.gradle.openapi.generator.settings.JsonSupport;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class NormalBuilderGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("allNecessityAndNullabilityVariants")
  void generate_when_allNecessityAndNullabilityVariants_then_correctOutput() {
    final NormalBuilderGenerator generator = new NormalBuilderGenerator();

    final Writer writer =
        generator.generate(
            ((JavaObjectPojo) JavaPojos.allNecessityAndNullabilityVariants())
                .getNormalBuilderContent(),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OPTIONAL::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));
    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_POJO_BUILDER::equals));
    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_PROPERTY::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void generate_when_additionalPropertyTypeIsList_then_containsListInRefs() {
    final NormalBuilderGenerator generator = new NormalBuilderGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.objectPojo(
                    PList.empty(), JavaAdditionalProperties.allowedFor(JavaTypes.STRING_LIST))
                .getNormalBuilderContent(),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_LIST::equals));
  }

  @Test
  @SnapshotName("allNecessityAndNullabilityVariantsDisabledJackson")
  void generate_when_allNecessityAndNullabilityVariantsDisabledJackson_then_correctOutput() {
    final NormalBuilderGenerator generator = new NormalBuilderGenerator();

    final Writer writer =
        generator.generate(
            ((JavaObjectPojo) JavaPojos.allNecessityAndNullabilityVariants())
                .getNormalBuilderContent(),
            TestPojoSettings.defaultSettings().withJsonSupport(JsonSupport.NONE),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OPTIONAL::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("allNecessityAndNullabilityVariantsDisabledSafeBuilder")
  void generate_when_allNecessityAndNullabilityVariantsDisabledSafeBuilder_then_correctOutput() {
    final NormalBuilderGenerator generator = new NormalBuilderGenerator();

    final Writer writer =
        generator.generate(
            ((JavaObjectPojo) JavaPojos.allNecessityAndNullabilityVariants())
                .getNormalBuilderContent(),
            TestPojoSettings.defaultSettings().withEnableSafeBuilder(false),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OPTIONAL::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));

    expect.toMatchSnapshot(writer.asString());
  }
}
