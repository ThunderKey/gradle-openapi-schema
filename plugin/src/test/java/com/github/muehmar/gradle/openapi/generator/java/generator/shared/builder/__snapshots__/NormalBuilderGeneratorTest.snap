allNecessityAndNullabilityVariants=[
@JsonPOJOBuilder(withPrefix = "set")
public static final class Builder {

  private Builder() {
  }

  private String requiredStringVal;
  private String requiredNullableStringVal;
  private boolean isRequiredNullableStringValPresent = false;
  private String optionalStringVal;
  private String optionalNullableStringVal;
  private boolean isOptionalNullableStringValNull = false;
  private Map<String, Object> additionalProperties = new HashMap<>();

  /**
   * RequiredStringVal
   */
  @JsonProperty("requiredStringVal")
  private Builder setRequiredStringVal(String requiredStringVal) {
    this.requiredStringVal = requiredStringVal;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  @JsonProperty("requiredNullableStringVal")
  private Builder setRequiredNullableStringVal(String requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal;
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  private Builder setRequiredNullableStringVal(Optional<String> requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal.orElse(null);
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * OptionalStringVal
   */
  @JsonProperty("optionalStringVal")
  public Builder setOptionalStringVal(String optionalStringVal) {
    this.optionalStringVal = optionalStringVal;
    return this;
  }

  /**
   * OptionalStringVal
   */
  public Builder setOptionalStringVal(Optional<String> optionalStringVal) {
    this.optionalStringVal = optionalStringVal.orElse(null);
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  @JsonProperty("optionalNullableStringVal")
  public Builder setOptionalNullableStringVal(String optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal;
    this.isOptionalNullableStringValNull = optionalNullableStringVal == null;
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  public Builder setOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal.onValue(val -> val).onNull(() -> null).onAbsent(() -> null);
    this.isOptionalNullableStringValNull = optionalNullableStringVal.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false);
    return this;
  }

  @JsonAnySetter
  public Builder addAdditionalProperty(String key, Object value) {
    this.additionalProperties.put(key, value);
    return this;
  }

  public Builder setAdditionalProperties(Map<String, Object> additionalProperties) {
    this.additionalProperties = new HashMap<>(additionalProperties);
    return this;
  }

  public NecessityAndNullabilityDto build() {
    return new NecessityAndNullabilityDto(requiredStringVal, requiredNullableStringVal, isRequiredNullableStringValPresent, optionalStringVal, optionalNullableStringVal, isOptionalNullableStringValNull);
  }
}
]


allNecessityAndNullabilityVariantsDisabledJackson=[
public static final class Builder {

  private Builder() {
  }

  private String requiredStringVal;
  private String requiredNullableStringVal;
  private boolean isRequiredNullableStringValPresent = false;
  private String optionalStringVal;
  private String optionalNullableStringVal;
  private boolean isOptionalNullableStringValNull = false;
  private Map<String, Object> additionalProperties = new HashMap<>();

  /**
   * RequiredStringVal
   */
  private Builder setRequiredStringVal(String requiredStringVal) {
    this.requiredStringVal = requiredStringVal;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  private Builder setRequiredNullableStringVal(String requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal;
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  private Builder setRequiredNullableStringVal(Optional<String> requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal.orElse(null);
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * OptionalStringVal
   */
  public Builder setOptionalStringVal(String optionalStringVal) {
    this.optionalStringVal = optionalStringVal;
    return this;
  }

  /**
   * OptionalStringVal
   */
  public Builder setOptionalStringVal(Optional<String> optionalStringVal) {
    this.optionalStringVal = optionalStringVal.orElse(null);
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  public Builder setOptionalNullableStringVal(String optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal;
    this.isOptionalNullableStringValNull = optionalNullableStringVal == null;
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  public Builder setOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal.onValue(val -> val).onNull(() -> null).onAbsent(() -> null);
    this.isOptionalNullableStringValNull = optionalNullableStringVal.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false);
    return this;
  }

  public Builder addAdditionalProperty(String key, Object value) {
    this.additionalProperties.put(key, value);
    return this;
  }

  public Builder setAdditionalProperties(Map<String, Object> additionalProperties) {
    this.additionalProperties = new HashMap<>(additionalProperties);
    return this;
  }

  public NecessityAndNullabilityDto build() {
    return new NecessityAndNullabilityDto(requiredStringVal, requiredNullableStringVal, isRequiredNullableStringValPresent, optionalStringVal, optionalNullableStringVal, isOptionalNullableStringValNull);
  }
}
]


allNecessityAndNullabilityVariantsDisabledSafeBuilder=[
public static Builder newBuilder() {
  return new Builder();
}

@JsonPOJOBuilder(withPrefix = "set")
public static final class Builder {

  private String requiredStringVal;
  private String requiredNullableStringVal;
  private boolean isRequiredNullableStringValPresent = false;
  private String optionalStringVal;
  private String optionalNullableStringVal;
  private boolean isOptionalNullableStringValNull = false;
  private Map<String, Object> additionalProperties = new HashMap<>();

  /**
   * RequiredStringVal
   */
  @JsonProperty("requiredStringVal")
  public Builder setRequiredStringVal(String requiredStringVal) {
    this.requiredStringVal = requiredStringVal;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  @JsonProperty("requiredNullableStringVal")
  public Builder setRequiredNullableStringVal(String requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal;
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * RequiredNullableStringVal
   */
  public Builder setRequiredNullableStringVal(Optional<String> requiredNullableStringVal) {
    this.requiredNullableStringVal = requiredNullableStringVal.orElse(null);
    this.isRequiredNullableStringValPresent = true;
    return this;
  }

  /**
   * OptionalStringVal
   */
  @JsonProperty("optionalStringVal")
  public Builder setOptionalStringVal(String optionalStringVal) {
    this.optionalStringVal = optionalStringVal;
    return this;
  }

  /**
   * OptionalStringVal
   */
  public Builder setOptionalStringVal(Optional<String> optionalStringVal) {
    this.optionalStringVal = optionalStringVal.orElse(null);
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  @JsonProperty("optionalNullableStringVal")
  public Builder setOptionalNullableStringVal(String optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal;
    this.isOptionalNullableStringValNull = optionalNullableStringVal == null;
    return this;
  }

  /**
   * OptionalNullableStringVal
   */
  public Builder setOptionalNullableStringVal(Tristate<String> optionalNullableStringVal) {
    this.optionalNullableStringVal = optionalNullableStringVal.onValue(val -> val).onNull(() -> null).onAbsent(() -> null);
    this.isOptionalNullableStringValNull = optionalNullableStringVal.onValue(ignore -> false).onNull(() -> true).onAbsent(() -> false);
    return this;
  }

  @JsonAnySetter
  public Builder addAdditionalProperty(String key, Object value) {
    this.additionalProperties.put(key, value);
    return this;
  }

  public Builder setAdditionalProperties(Map<String, Object> additionalProperties) {
    this.additionalProperties = new HashMap<>(additionalProperties);
    return this;
  }

  public NecessityAndNullabilityDto build() {
    return new NecessityAndNullabilityDto(requiredStringVal, requiredNullableStringVal, isRequiredNullableStringValPresent, optionalStringVal, optionalNullableStringVal, isOptionalNullableStringValNull);
  }
}
]